src/proofs/proof.rs

```diff
@@ -0,0 +1,514 @@
+use symbol_table::GlobalSymbol as Symbol;
+
+use crate::{
+    ast::{Command, Expr, Fact, GenericAction, GenericFact, Rule},
+    HashMap, HashSet, Term, TermDag,
+};
+
+type ProofId = usize;
+
+#[derive(Clone)]
+pub struct ProofStore {
+    store: Vec<ProofTerm>,
+    memo: HashMap<ProofTerm, ProofId>,
+    termdag: TermDag,
+}
+
+type Substitution = Vec<(Symbol, Term)>;
+
+fn subst_get(subst: &Substitution, sym: Symbol) -> Option<&Term> {
+    for ele in subst {
+        if ele.0 == sym {
+            return Some(&ele.1);
+        }
+    }
+    None
+}
+
+#[derive(Clone, PartialEq, Eq, Hash, Debug)]
+pub enum Proposition {
+    TOk(Term),
+    TEq(Term, Term),
+}
+
+// todo how to ignore this warning?
+#[warn(clippy::enum_variant_names)]
+#[derive(Clone, PartialEq, Eq, Hash)]
+pub enum ProofTerm {
+    /// proves a Proposition based on a rule application
+    /// the subsitution gives the mapping from variables to terms
+    /// the body_pfs gives proofs for each of the conditions in the query of the rule
+    /// the act_pf gives a location in the action of the proposition
+    PRule {
+        rule_name: Symbol,
+        subst: Substitution,
+        body_pfs: Vec<ProofId>,
```

Comment:
What do these "body_pfs" mean? are they premises? Will they be a bunch of equality proofs or TOK proofs?

Reply:
They are proofs for each line of the rule's premises
They produces a `Proposition` that matches the rule's syntax (a TOK proof for an expression and a equality proof for an (= t1 t2) line)